BioInformatics II
CSCI 6595
Spring 2013

* No data or results included in the repository.

Assignment 1:
A program that opens and parses FASTA files, identifying the type (DNA/RNA/Protein) of
sequences and calculates the relative frequency for each acid.  

How to run:
1) Navigate to the directory containing BioProgram.jar
2) type: java -jar BioProgram.jar

NOTES:
*  Large data files (300+ MBs) have not been included for Vitis vinifera and Nicotiana Tabacum.  
   Smaller samples have been provided, as well as the results from the complete data files.
   
Assignment 2:
Results can be found in:

data/Mus Xenopus alignment.txt
data/Hyphantria Anopheles alignment.txt

You can also run the TUI using BioProgram.jar.  
Select option 3 for Sequence Alignments.  Choose any option,
the alignments will be displayed to the screen and written
to a file.  

Assignment 3:
Hidden Markov Models

You can run the TUI in bin\BioProgram.jar.  

You have a few options to work with HMMs.
1) A complete TUI system for designing HMMs (either from
   scratch or from the two default models) has been provided.
   Reports can be built and written to file for the different
   decodings.
2) A number of "default" trials have been provided
   which run default conditions.
   ** The actuals testing mechanism for the report is provided
      in this menu.  It automatically saves reports to 
	  default files in /results.  